src/sidebar/components/annotation-editor.js

```diff
@@ -0,0 +1,181 @@
+import { createElement } from 'preact';
+import { useRef } from 'preact/hooks';
+import propTypes from 'prop-types';
+
+import { normalizeKeyName } from '../../shared/browser-compatibility-utils';
+import { withServices } from '../util/service-context';
+import { applyTheme } from '../util/theme';
+import useStore from '../store/use-store';
+
+import AnnotationLicense from './annotation-license';
+import AnnotationPublishControl from './annotation-publish-control';
+import MarkdownEditor from './markdown-editor';
+import TagEditor from './tag-editor';
+
+/**
+ * @typedef {import("../../types/api").Annotation} Annotation
+ * @typedef {import("../../types/config").MergedConfig} MergedConfig
+ */
+
+/**
+ * @typedef AnnotationEditorProps
+ * @prop {Annotation} annotation - The annotation under edit
+ * @prop {Object} annotationsService - Injected service
+ * @prop {MergedConfig} settings - Injected service
+ * @prop {Object} toastMessenger - Injected service
+ * @prop {Object} tags - Injected service
+ */
+
+/**
+ * Display annotation content in an editable format.
+ *
+ * @param {AnnotationEditorProps} props
+ */
+function AnnotationEditor({
+  annotation,
+  annotationsService,
+  settings,
+  tags: tagsService,
+  toastMessenger,
+}) {
+  // Create a ref to the `input` element for editing tags
+  const tagInputRef = useRef(/** @type {HTMLInputElement|null} */ (null));
```

Comment:
Just a suggestion - an alternative to exposing the `<input>` from the `TagEditor` directly would be to expose only the pending tag via a callback that gets invoked each time the user enters text. This change wouldn't have a huge impact at present, but might make future changes easier by avoiding other components depending on TagEditor's internal elements.

<details>
<summary>Code sketch</summary>


```diff
diff --git a/src/sidebar/components/annotation-editor.js b/src/sidebar/components/annotation-editor.js
index 885daf959..4346705e0 100644
--- a/src/sidebar/components/annotation-editor.js
+++ b/src/sidebar/components/annotation-editor.js
@@ -1,5 +1,5 @@
 import { createElement } from 'preact';
-import { useRef } from 'preact/hooks';
+import { useState } from 'preact/hooks';
 import propTypes from 'prop-types';
 
 import { normalizeKeyName } from '../../shared/browser-compatibility-utils';
@@ -38,8 +38,7 @@ function AnnotationEditor({
   tags: tagsService,
   toastMessenger,
 }) {
-  // Create a ref to the `input` element for editing tags
-  const tagInputRef = useRef(/** @type {HTMLInputElement|null} */ (null));
+  const [pendingTag, setPendingTag] = useState(/** @type {string|null} */ (null));
 
   const draft =
     /** @type {Object|null} */
@@ -111,8 +110,8 @@ function AnnotationEditor({
     // If there is any content in the tag editor input field that has
     // not been committed as a tag, go ahead and add it as a tag
     // See https://github.com/hypothesis/product-backlog/issues/1122
-    if (tagInputRef?.current?.value) {
-      onAddTag(tagInputRef.current.value);
+    if (pendingTag) {
+      onAddTag(pendingTag);
     }
     try {
       await annotationsService.save(annotation);
@@ -148,7 +147,7 @@ function AnnotationEditor({
       <TagEditor
         onAddTag={onAddTag}
         onRemoveTag={onRemoveTag}
-        tagInputRef={tagInputRef}
+        onTagInput={setPendingTag}
         tagList={tags}
       />
       <div className="annotation__form-actions u-layout-row">
diff --git a/src/sidebar/components/tag-editor.js b/src/sidebar/components/tag-editor.js
index 5b2bd7b2d..09c80c439 100644
--- a/src/sidebar/components/tag-editor.js
+++ b/src/sidebar/components/tag-editor.js
@@ -19,7 +19,8 @@ let tagEditorIdCounter = 0;
  * @typedef TagEditorProps
  * @prop {(a: string) => boolean} onAddTag - Callback to add a tag to the annotation
  * @prop {(a: string) => boolean} onRemoveTag - Callback to remove a tag from the annotation
- * @prop {Object} tagInputRef - A reference for the `input` field
+ * @prop {(tag: string) => any} [onTagInput] -
+ *   Callback invoked with pending tag as the user types in the input field
  * @prop {string[]} tagList - The list of editable tags
  * @prop {Object} tags - Injected service
  */
@@ -35,10 +36,11 @@ let tagEditorIdCounter = 0;
 function TagEditor({
   onAddTag,
   onRemoveTag,
-  tagInputRef: inputEl,
+  onTagInput,
   tagList,
   tags: tagsService,
 }) {
+  const inputEl = useRef(/** @type {HTMLInputElement|null} */ (null));
   const [suggestions, setSuggestions] = useState(/** @type {string[]} */ ([]));
   const [activeItem, setActiveItem] = useState(-1); // -1 is unselected
   const [suggestionsListOpen, setSuggestionsListOpen] = useState(false);
@@ -117,6 +119,8 @@ function TagEditor({
    * @param {import("preact").JSX.TargetedEvent<HTMLInputElement, InputEvent>} e
    */
   const handleOnInput = e => {
+    onTagInput?.(inputEl.current.value);
+
     if (
       e.inputType === 'insertText' ||
       e.inputType === 'deleteContentBackward' ||
@@ -331,7 +335,7 @@ function TagEditor({
 TagEditor.propTypes = {
   onAddTag: propTypes.func.isRequired,
   onRemoveTag: propTypes.func.isRequired,
-  tagInputRef: propTypes.object.isRequired,
+  onTagInput: propTypes.func,
   tagList: propTypes.array.isRequired,
   tags: propTypes.object.isRequired,
 };
```
</details>

Reply:
I do like this pattern better. It exposes the piece of this we care about, not the details of an input field.

Reply:
However...does this mean `AnnotationEditor` will re-render on every keystroke?